sklearn/cluster/tests/test_dbscan.py

```diff
@@ -122,6 +122,27 @@ def test_dbscan_input_not_modified(metric, csr_container):
         assert_array_equal(X, X_copy)
 
 
+@pytest.mark.parametrize("csr_container", CSR_CONTAINERS)
+def test_dbscan_input_not_modified_precomputed_sparse_nodiag(csr_container):
+    """Check that we don't modify in-place the pre-computed sparse matrix.
+
+    Non-regression test for:
+    https://github.com/scikit-learn/scikit-learn/issues/27508
+    """
+    X = np.random.RandomState(0).rand(10, 10)
+    # Add zeros on the diagonal that will be implicit when creating
+    # the sparse matrix. If `X` is modified in-place, the zeros from
+    # the diagonal will be made explicit.
+    np.fill_diagonal(X, 0)
+    X = csr_container(X)
```

Comment:
We could make an assert just below this line to be sure that we have implicit zeros.